client/src/controllers/KeyboardController.test.js

```diff
@@ -0,0 +1,67 @@
+import { Application } from '@hotwired/stimulus';
+import { KeyboardController } from './KeyboardController';
+
+describe('KeyboardController', () => {
+  let app;
+  const oldWindowLocation = window.location;
```

Comment:
No need to do anything with mocking the location in this test, that's just to support a different test though.

I have run your code locally and it's super close, but here's a refined way to test for the keyboard trigger.

```suggestion
  const buttonClickMock = jest.fn();

  /**
   * Simulates a keydown, keypress, and keyup event for the provided key.
   */
  const simulateKey = (
    { keyCode, which = keyCode.charCodeAt(0) },
    target = document.body,
  ) =>
    Object.fromEntries(
      ['keydown', 'keypress', 'keyup'].map((type) => {
        const defaultPrevented = target.dispatchEvent(
          new KeyboardEvent(type, {
            bubbles: true,
            cancelable: true,
            keyCode,
            which,
          }),
        );
        return [type, defaultPrevented];
      }),
    );
```